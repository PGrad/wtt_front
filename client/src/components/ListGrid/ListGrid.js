import React, { useState } from 'react';
import './ListGrid.scss';
import { Edit, Verified, GppBad } from '@mui/icons-material';

function sortList(data, sortColumn, sortOrderAsc = true) {
  return [...data].sort((a, b) => {
    const aa = a[sortColumn];
    const bb = b[sortColumn];

    if (aa === bb) return 0;
    if (aa === null || aa === undefined) return sortOrderAsc ? 1 : -1;
    if (bb === null || bb === undefined) return sortOrderAsc ? -1 : 1;

    // handle boolean value sorting for the "broken" column in the Source page
    if (typeof aa === 'boolean' && typeof bb === 'boolean') {
      return sortOrderAsc ? aa - bb : bb - aa;
    }

    const aStr = String(aa).toLowerCase();
    const bStr = String(bb).toLowerCase();

    if (sortOrderAsc) {
      // If ascending, compare the strings and return either -1 or 1 based on the result.
      return aStr < bStr ? -1 : 1;
    } else {
      // If descending, compare the strings and return either -1 or 1 based on the opposite result.
      return aStr > bStr ? -1 : 1;
    }
  });
}

function createDynamicGridTemplateColumns(columnsNumber, listType) {
  // the case of listType source is used for the Source page to create specific column widths
  // the default case is used for the Data Page (or any other page that uses ListGrid) to create equal column widths
  switch (listType) {
    case 'source':
      return '2em 2.3em 8em 3.5em 3em 1fr 1fr 8.5em 1fr';
    default: {
      let templateColumns = '';
      for (let i = 0; i < columnsNumber; i++) {
        templateColumns += '1fr ';
      }
      return templateColumns;
    }
  }
}

export function ListGrid({
  data,
  columns,
  openEdit,
  setOpenEdit,
  defaultSortIndex = 0,
  setList,
  listType,
}) {
  const defaultSortColumn = columns[defaultSortIndex].columnHeader;
  const [sortOrderAsc, setSortOrderAsc] = useState(true);
  const [lastSortColumn, setLastSortColumn] = useState(defaultSortColumn);
  let sortColumn = lastSortColumn;

  if (!columns.find(({ columnHeader }) => columnHeader === sortColumn)) {
    // The new data doesn't include the last column we sorted on, so default to common/asc and
    // update the state for the next render.
    sortColumn = defaultSortColumn;
    setLastSortColumn(sortColumn);
    setSortOrderAsc(true);
  }

  const sortedList = sortList(data, sortColumn, sortOrderAsc);

  const handleSort = (event) => {
    const newColumn = event.target.value;
    let newOrder = !sortOrderAsc;
    if (newColumn !== sortColumn) {
      // Reset the sort order to ascending when the sort column changes.
      newOrder = true;
    }
    setSortOrderAsc(newOrder);
    setLastSortColumn(newColumn);
  };

  return (
    <div className="listgrid__list" key={sortedList[0]}>
      <ListGridColumnHeaders
        handleSort={handleSort}
        columns={columns}
        listType={listType}
      />
      {sortedList.map((item) => (
        <ListGridItem
          item={item}
          columns={columns}
          key={item.id}
          openEdit={openEdit}
          setOpenEdit={setOpenEdit}
          setList={setList}
          listType={listType}
        />
      ))}
    </div>
  );
}

function ListGridColumnHeaders({ handleSort, columns, listType }) {
  return (
    <div
      className="listgrid__columnheader"
      style={{
        gridTemplateColumns: createDynamicGridTemplateColumns(
          columns.length,
          listType,
        ),
      }}
    >
      {columns.map(({ columnHeader, label }) => (
        <div className="listgrid__columnheader-item" key={columnHeader}>
          <button
            type="button"
            className="listgrid__columnheader-btn"
            value={columnHeader}
            onClick={handleSort}
          >
            {label}
          </button>
        </div>
      ))}
    </div>
  );
}

function createKey(item, columnHeader, index) {
  return item[columnHeader]
    ? `${columnHeader}-${item[columnHeader]}`
    : `${columnHeader}-${index}`;
}

function ListGridItem({ item, columns, setOpenEdit, setList, listType }) {
  return (
    <div
      className="listgrid__list-item"
      style={{
        gridTemplateColumns: createDynamicGridTemplateColumns(
          columns.length,
          listType,
        ),
      }}
    >
      {columns.map(({ columnHeader }, index) => (
        <div
          className="listgrid__list-item-item"
          key={createKey(item, columnHeader, index)}
        >
          <ListGridItemSwitch
            item={item}
            columnHeader={columnHeader}
            setOpenEdit={setOpenEdit}
            setList={setList}
          />
        </div>
      ))}
    </div>
  );
}

function ListGridItemSwitch({ item, columnHeader, setOpenEdit, setList }) {
  switch (columnHeader) {
    case 'idSourceName':
      return (
        <ListGridIdSourceName
          source={item}
          setOpenEdit={setOpenEdit}
          setList={setList}
        />
      );
    case 'city':
      return (
        <ListGridItemDownload
          downloadUrl={item.download}
          itemDownloadName={item.city}
          setOpenEdit={setOpenEdit}
        />
      );
    case 'broken':
      return <ListGridItemCheckbox checked={item.broken} />;
    case 'email':
      return <ListGridItemEmail email={item.email} broken={item.broken} />;
    default:
      return item[columnHeader];
  }
}

function ListGridIdSourceName({ source, setOpenEdit, setList }) {
  const handleClick = () => {
    setOpenEdit(true);
    setList(source);
  };
  return (
    <div
      role="button"
      className="listgrid__list-item-editbutton"
      aria-label="Edit"
      tabIndex="0"
      onKeyDown={handleClick}
      onClick={handleClick}
    >
      <Edit color="info" fontSize="small" />
    </div>
  );
}

function ListGridItemDownload({ downloadUrl, itemDownloadName }) {
  return (
    <span className="listgrid__download">
      <a href={downloadUrl} target="_blank" rel="noreferrer">
        {itemDownloadName}
      </a>
    </span>
  );
}

function createEmailLink(email, broken) {
  let subject, body;

  if (broken) {
    subject = 'Water the Trees Source update request';
    body = `Hi,\n The tree source link for your city is broken, can you please update it here ${encodeURIComponent(
      'https://waterthetrees.com/data',
    )}`;
  } else {
    subject = 'Water the Trees Source inquiry';
    body = `Hi,\n I have a question about your tree data on ${encodeURIComponent(
      'https://waterthetrees.com/data',
    )}.`;
  }

  return `mailto:${email}?subject=${encodeURIComponent(
    subject,
  )}&body=${encodeURIComponent(body)}`;
}

function ListGridItemEmail({ email, broken }) {
  const mailTo = createEmailLink(email, broken);
  return (
    <a href={mailTo} target="_blank" rel="noreferrer">
      {email}
    </a>
  );
}

function ListGridItemCheckbox({ checked }) {
  return (
    <React.Fragment>
      {checked ? <GppBad color="error" /> : <Verified color="success" />}
    </React.Fragment>
  );
}
