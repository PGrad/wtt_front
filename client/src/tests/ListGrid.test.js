import React from 'react';
import { render } from '@testing-library/react';
import { ListGrid } from '@/components/ListGrid/ListGrid';

/**
 * These snapshots are designed to catch _any_ change to UI components, so they
 * are by definition very brittle. If the changes are related to your changes
 * and look expected, update these snapshots with `npm test -- -u`
 */

const mockSetOpenEdit = jest.fn(() => {});
const mockSetList = jest.fn(() => {});
const mockHandleNewEditSource = jest.fn(() => {});

function createColumnHeaders(source) {
  if (!source) return [];
  return [
    {
      columnHeader: 'idSourceName',
      label: '',
    },
    {
      columnHeader: 'broken',
      label: 'Up',
    },
    {
      columnHeader: 'city',
      label: 'Location',
    },
    {
      columnHeader: 'isoAlpha3',
      label: 'Country',
    },
    {
      columnHeader: 'state',
      label: 'State',
    },

    {
      columnHeader: 'email',
      label: 'Email',
    },
    {
      columnHeader: 'contact',
      label: 'Contact',
    },
    {
      columnHeader: 'phone',
      label: 'Phone',
    },

    {
      columnHeader: 'notes',
      label: 'Notes',
    },
  ];
}

const mockSources = [
  {
    idSourceName: 'rochester',
    isoAlpha3: 'USA',
    country: 'United States of America',
    state: 'NY',
    city: 'Rochester',
    email: null,
    contact: null,
    phone: null,
    info: 'http://hub.arcgis.com/datasets/RochesterNY::trees-open-data',
    download:
      'https://opendata.arcgis.com/datasets/4c209944e2984b4a908a14b0cbe48075_0.zip',
    notes: null,
    filename: null,
    format: 'zip',
    longitude: null,
    latitude: null,
    license: null,
    broken: true,
  },
  {
    idSourceName: 'richardson',
    isoAlpha3: 'USA',
    country: 'United States of America',
    state: 'TX',
    city: 'Richardson',
    email: null,
    contact: null,
    phone: null,
    info: 'http://hub.arcgis.com/datasets/richardson::trees',
    download:
      'https://opendata.arcgis.com/datasets/cd10a9e85354488dbdb697ce97ccb064_0.csv',
    notes: null,
    filename: null,
    format: null,
    longitude: null,
    latitude: null,
    license: null,
    broken: false,
  },
];

describe('<Source /> spec', () => {
  it('renders ListGrid correctly', () => {
    const columns = createColumnHeaders(mockSources);
    const listGrid = render(
      <ListGrid
        data={mockSources}
        columns={columns}
        defaultSortIndex={0}
        handleNewEditSource={mockHandleNewEditSource}
        listType="source"
        setOpenEdit={mockSetOpenEdit}
        setList={mockSetList}
        openEdit={false}
      />,
    );
    expect(listGrid).toMatchSnapshot();
  });
});
