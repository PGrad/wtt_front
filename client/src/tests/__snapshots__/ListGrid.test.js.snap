// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Source /> spec renders ListGrid correctly 1`] = `
{
  "asFragment": [Function],
  "baseElement": .emotion-0 {
  -webkit-user-select: none;
  -moz-user-select: none;
  -ms-user-select: none;
  user-select: none;
  width: 1em;
  height: 1em;
  display: inline-block;
  fill: currentColor;
  -webkit-flex-shrink: 0;
  -ms-flex-negative: 0;
  flex-shrink: 0;
  -webkit-transition: fill 200ms cubic-bezier(0.4, 0, 0.2, 1) 0ms;
  transition: fill 200ms cubic-bezier(0.4, 0, 0.2, 1) 0ms;
  font-size: 1.25rem;
  color: #0288d1;
}

.emotion-1 {
  -webkit-user-select: none;
  -moz-user-select: none;
  -ms-user-select: none;
  user-select: none;
  width: 1em;
  height: 1em;
  display: inline-block;
  fill: currentColor;
  -webkit-flex-shrink: 0;
  -ms-flex-negative: 0;
  flex-shrink: 0;
  -webkit-transition: fill 200ms cubic-bezier(0.4, 0, 0.2, 1) 0ms;
  transition: fill 200ms cubic-bezier(0.4, 0, 0.2, 1) 0ms;
  font-size: 1.5rem;
  color: #2e7d32;
}

.emotion-3 {
  -webkit-user-select: none;
  -moz-user-select: none;
  -ms-user-select: none;
  user-select: none;
  width: 1em;
  height: 1em;
  display: inline-block;
  fill: currentColor;
  -webkit-flex-shrink: 0;
  -ms-flex-negative: 0;
  flex-shrink: 0;
  -webkit-transition: fill 200ms cubic-bezier(0.4, 0, 0.2, 1) 0ms;
  transition: fill 200ms cubic-bezier(0.4, 0, 0.2, 1) 0ms;
  font-size: 1.5rem;
  color: #d32f2f;
}

<body>
    <div>
      <div
        class="listgrid__list"
      >
        <div
          class="listgrid__columnheader"
          style="grid-template-columns: 2em 2.3em 8em 3.5em 3em 1fr 1fr 8.5em 1fr;"
        >
          <div
            class="listgrid__columnheader-item"
          >
            <button
              class="listgrid__columnheader-btn"
              type="button"
              value="idSourceName"
            />
          </div>
          <div
            class="listgrid__columnheader-item"
          >
            <button
              class="listgrid__columnheader-btn"
              type="button"
              value="broken"
            >
              Up
            </button>
          </div>
          <div
            class="listgrid__columnheader-item"
          >
            <button
              class="listgrid__columnheader-btn"
              type="button"
              value="city"
            >
              Location
            </button>
          </div>
          <div
            class="listgrid__columnheader-item"
          >
            <button
              class="listgrid__columnheader-btn"
              type="button"
              value="isoAlpha3"
            >
              Country
            </button>
          </div>
          <div
            class="listgrid__columnheader-item"
          >
            <button
              class="listgrid__columnheader-btn"
              type="button"
              value="state"
            >
              State
            </button>
          </div>
          <div
            class="listgrid__columnheader-item"
          >
            <button
              class="listgrid__columnheader-btn"
              type="button"
              value="email"
            >
              Email
            </button>
          </div>
          <div
            class="listgrid__columnheader-item"
          >
            <button
              class="listgrid__columnheader-btn"
              type="button"
              value="contact"
            >
              Contact
            </button>
          </div>
          <div
            class="listgrid__columnheader-item"
          >
            <button
              class="listgrid__columnheader-btn"
              type="button"
              value="phone"
            >
              Phone
            </button>
          </div>
          <div
            class="listgrid__columnheader-item"
          >
            <button
              class="listgrid__columnheader-btn"
              type="button"
              value="notes"
            >
              Notes
            </button>
          </div>
        </div>
        <div
          class="listgrid__list-item"
          style="grid-template-columns: 2em 2.3em 8em 3.5em 3em 1fr 1fr 8.5em 1fr;"
        >
          <div
            class="listgrid__list-item-item"
          >
            <div
              aria-label="Edit"
              class="listgrid__list-item-editbutton"
              role="button"
              tabindex="0"
            >
              <svg
                aria-hidden="true"
                class="MuiSvgIcon-root MuiSvgIcon-colorInfo MuiSvgIcon-fontSizeSmall emotion-0"
                data-testid="EditIcon"
                focusable="false"
                viewBox="0 0 24 24"
              >
                <path
                  d="M3 17.25V21h3.75L17.81 9.94l-3.75-3.75L3 17.25zM20.71 7.04c.39-.39.39-1.02 0-1.41l-2.34-2.34a.9959.9959 0 0 0-1.41 0l-1.83 1.83 3.75 3.75 1.83-1.83z"
                />
              </svg>
            </div>
          </div>
          <div
            class="listgrid__list-item-item"
          >
            <svg
              aria-hidden="true"
              class="MuiSvgIcon-root MuiSvgIcon-colorSuccess MuiSvgIcon-fontSizeMedium emotion-1"
              data-testid="VerifiedIcon"
              focusable="false"
              viewBox="0 0 24 24"
            >
              <path
                d="m23 12-2.44-2.79.34-3.69-3.61-.82-1.89-3.2L12 2.96 8.6 1.5 6.71 4.69 3.1 5.5l.34 3.7L1 12l2.44 2.79-.34 3.7 3.61.82L8.6 22.5l3.4-1.47 3.4 1.46 1.89-3.19 3.61-.82-.34-3.69L23 12zm-12.91 4.72-3.8-3.81 1.48-1.48 2.32 2.33 5.85-5.87 1.48 1.48-7.33 7.35z"
              />
            </svg>
          </div>
          <div
            class="listgrid__list-item-item"
          >
            <span
              class="listgrid__download"
            >
              <a
                href="https://opendata.arcgis.com/datasets/cd10a9e85354488dbdb697ce97ccb064_0.csv"
                rel="noreferrer"
                target="_blank"
              >
                Richardson
              </a>
            </span>
          </div>
          <div
            class="listgrid__list-item-item"
          >
            USA
          </div>
          <div
            class="listgrid__list-item-item"
          >
            TX
          </div>
          <div
            class="listgrid__list-item-item"
          >
            <a
              href="mailto:null?subject=Water%20the%20Trees%20Source%20inquiry&body=Hi%2C%0A%20I%20have%20a%20question%20about%20your%20tree%20data%20on%20https%253A%252F%252Fwaterthetrees.com%252Fdata."
              rel="noreferrer"
              target="_blank"
            />
          </div>
          <div
            class="listgrid__list-item-item"
          />
          <div
            class="listgrid__list-item-item"
          />
          <div
            class="listgrid__list-item-item"
          />
        </div>
        <div
          class="listgrid__list-item"
          style="grid-template-columns: 2em 2.3em 8em 3.5em 3em 1fr 1fr 8.5em 1fr;"
        >
          <div
            class="listgrid__list-item-item"
          >
            <div
              aria-label="Edit"
              class="listgrid__list-item-editbutton"
              role="button"
              tabindex="0"
            >
              <svg
                aria-hidden="true"
                class="MuiSvgIcon-root MuiSvgIcon-colorInfo MuiSvgIcon-fontSizeSmall emotion-0"
                data-testid="EditIcon"
                focusable="false"
                viewBox="0 0 24 24"
              >
                <path
                  d="M3 17.25V21h3.75L17.81 9.94l-3.75-3.75L3 17.25zM20.71 7.04c.39-.39.39-1.02 0-1.41l-2.34-2.34a.9959.9959 0 0 0-1.41 0l-1.83 1.83 3.75 3.75 1.83-1.83z"
                />
              </svg>
            </div>
          </div>
          <div
            class="listgrid__list-item-item"
          >
            <svg
              aria-hidden="true"
              class="MuiSvgIcon-root MuiSvgIcon-colorError MuiSvgIcon-fontSizeMedium emotion-3"
              data-testid="GppBadIcon"
              focusable="false"
              viewBox="0 0 24 24"
            >
              <path
                d="M12 2 4 5v6.09c0 5.05 3.41 9.76 8 10.91 4.59-1.15 8-5.86 8-10.91V5l-8-3zm3.5 12.09-1.41 1.41L12 13.42 9.91 15.5 8.5 14.09 10.59 12 8.5 9.91 9.91 8.5 12 10.59l2.09-2.09 1.41 1.41L13.42 12l2.08 2.09z"
              />
            </svg>
          </div>
          <div
            class="listgrid__list-item-item"
          >
            <span
              class="listgrid__download"
            >
              <a
                href="https://opendata.arcgis.com/datasets/4c209944e2984b4a908a14b0cbe48075_0.zip"
                rel="noreferrer"
                target="_blank"
              >
                Rochester
              </a>
            </span>
          </div>
          <div
            class="listgrid__list-item-item"
          >
            USA
          </div>
          <div
            class="listgrid__list-item-item"
          >
            NY
          </div>
          <div
            class="listgrid__list-item-item"
          >
            <a
              href="mailto:null?subject=Water%20the%20Trees%20Source%20update%20request&body=Hi%2C%0A%20The%20tree%20source%20link%20for%20your%20city%20is%20broken%2C%20can%20you%20please%20update%20it%20here%20https%253A%252F%252Fwaterthetrees.com%252Fdata"
              rel="noreferrer"
              target="_blank"
            />
          </div>
          <div
            class="listgrid__list-item-item"
          />
          <div
            class="listgrid__list-item-item"
          />
          <div
            class="listgrid__list-item-item"
          />
        </div>
      </div>
    </div>
  </body>,
  "container": .emotion-0 {
  -webkit-user-select: none;
  -moz-user-select: none;
  -ms-user-select: none;
  user-select: none;
  width: 1em;
  height: 1em;
  display: inline-block;
  fill: currentColor;
  -webkit-flex-shrink: 0;
  -ms-flex-negative: 0;
  flex-shrink: 0;
  -webkit-transition: fill 200ms cubic-bezier(0.4, 0, 0.2, 1) 0ms;
  transition: fill 200ms cubic-bezier(0.4, 0, 0.2, 1) 0ms;
  font-size: 1.25rem;
  color: #0288d1;
}

.emotion-1 {
  -webkit-user-select: none;
  -moz-user-select: none;
  -ms-user-select: none;
  user-select: none;
  width: 1em;
  height: 1em;
  display: inline-block;
  fill: currentColor;
  -webkit-flex-shrink: 0;
  -ms-flex-negative: 0;
  flex-shrink: 0;
  -webkit-transition: fill 200ms cubic-bezier(0.4, 0, 0.2, 1) 0ms;
  transition: fill 200ms cubic-bezier(0.4, 0, 0.2, 1) 0ms;
  font-size: 1.5rem;
  color: #2e7d32;
}

.emotion-3 {
  -webkit-user-select: none;
  -moz-user-select: none;
  -ms-user-select: none;
  user-select: none;
  width: 1em;
  height: 1em;
  display: inline-block;
  fill: currentColor;
  -webkit-flex-shrink: 0;
  -ms-flex-negative: 0;
  flex-shrink: 0;
  -webkit-transition: fill 200ms cubic-bezier(0.4, 0, 0.2, 1) 0ms;
  transition: fill 200ms cubic-bezier(0.4, 0, 0.2, 1) 0ms;
  font-size: 1.5rem;
  color: #d32f2f;
}

<div>
    <div
      class="listgrid__list"
    >
      <div
        class="listgrid__columnheader"
        style="grid-template-columns: 2em 2.3em 8em 3.5em 3em 1fr 1fr 8.5em 1fr;"
      >
        <div
          class="listgrid__columnheader-item"
        >
          <button
            class="listgrid__columnheader-btn"
            type="button"
            value="idSourceName"
          />
        </div>
        <div
          class="listgrid__columnheader-item"
        >
          <button
            class="listgrid__columnheader-btn"
            type="button"
            value="broken"
          >
            Up
          </button>
        </div>
        <div
          class="listgrid__columnheader-item"
        >
          <button
            class="listgrid__columnheader-btn"
            type="button"
            value="city"
          >
            Location
          </button>
        </div>
        <div
          class="listgrid__columnheader-item"
        >
          <button
            class="listgrid__columnheader-btn"
            type="button"
            value="isoAlpha3"
          >
            Country
          </button>
        </div>
        <div
          class="listgrid__columnheader-item"
        >
          <button
            class="listgrid__columnheader-btn"
            type="button"
            value="state"
          >
            State
          </button>
        </div>
        <div
          class="listgrid__columnheader-item"
        >
          <button
            class="listgrid__columnheader-btn"
            type="button"
            value="email"
          >
            Email
          </button>
        </div>
        <div
          class="listgrid__columnheader-item"
        >
          <button
            class="listgrid__columnheader-btn"
            type="button"
            value="contact"
          >
            Contact
          </button>
        </div>
        <div
          class="listgrid__columnheader-item"
        >
          <button
            class="listgrid__columnheader-btn"
            type="button"
            value="phone"
          >
            Phone
          </button>
        </div>
        <div
          class="listgrid__columnheader-item"
        >
          <button
            class="listgrid__columnheader-btn"
            type="button"
            value="notes"
          >
            Notes
          </button>
        </div>
      </div>
      <div
        class="listgrid__list-item"
        style="grid-template-columns: 2em 2.3em 8em 3.5em 3em 1fr 1fr 8.5em 1fr;"
      >
        <div
          class="listgrid__list-item-item"
        >
          <div
            aria-label="Edit"
            class="listgrid__list-item-editbutton"
            role="button"
            tabindex="0"
          >
            <svg
              aria-hidden="true"
              class="MuiSvgIcon-root MuiSvgIcon-colorInfo MuiSvgIcon-fontSizeSmall emotion-0"
              data-testid="EditIcon"
              focusable="false"
              viewBox="0 0 24 24"
            >
              <path
                d="M3 17.25V21h3.75L17.81 9.94l-3.75-3.75L3 17.25zM20.71 7.04c.39-.39.39-1.02 0-1.41l-2.34-2.34a.9959.9959 0 0 0-1.41 0l-1.83 1.83 3.75 3.75 1.83-1.83z"
              />
            </svg>
          </div>
        </div>
        <div
          class="listgrid__list-item-item"
        >
          <svg
            aria-hidden="true"
            class="MuiSvgIcon-root MuiSvgIcon-colorSuccess MuiSvgIcon-fontSizeMedium emotion-1"
            data-testid="VerifiedIcon"
            focusable="false"
            viewBox="0 0 24 24"
          >
            <path
              d="m23 12-2.44-2.79.34-3.69-3.61-.82-1.89-3.2L12 2.96 8.6 1.5 6.71 4.69 3.1 5.5l.34 3.7L1 12l2.44 2.79-.34 3.7 3.61.82L8.6 22.5l3.4-1.47 3.4 1.46 1.89-3.19 3.61-.82-.34-3.69L23 12zm-12.91 4.72-3.8-3.81 1.48-1.48 2.32 2.33 5.85-5.87 1.48 1.48-7.33 7.35z"
            />
          </svg>
        </div>
        <div
          class="listgrid__list-item-item"
        >
          <span
            class="listgrid__download"
          >
            <a
              href="https://opendata.arcgis.com/datasets/cd10a9e85354488dbdb697ce97ccb064_0.csv"
              rel="noreferrer"
              target="_blank"
            >
              Richardson
            </a>
          </span>
        </div>
        <div
          class="listgrid__list-item-item"
        >
          USA
        </div>
        <div
          class="listgrid__list-item-item"
        >
          TX
        </div>
        <div
          class="listgrid__list-item-item"
        >
          <a
            href="mailto:null?subject=Water%20the%20Trees%20Source%20inquiry&body=Hi%2C%0A%20I%20have%20a%20question%20about%20your%20tree%20data%20on%20https%253A%252F%252Fwaterthetrees.com%252Fdata."
            rel="noreferrer"
            target="_blank"
          />
        </div>
        <div
          class="listgrid__list-item-item"
        />
        <div
          class="listgrid__list-item-item"
        />
        <div
          class="listgrid__list-item-item"
        />
      </div>
      <div
        class="listgrid__list-item"
        style="grid-template-columns: 2em 2.3em 8em 3.5em 3em 1fr 1fr 8.5em 1fr;"
      >
        <div
          class="listgrid__list-item-item"
        >
          <div
            aria-label="Edit"
            class="listgrid__list-item-editbutton"
            role="button"
            tabindex="0"
          >
            <svg
              aria-hidden="true"
              class="MuiSvgIcon-root MuiSvgIcon-colorInfo MuiSvgIcon-fontSizeSmall emotion-0"
              data-testid="EditIcon"
              focusable="false"
              viewBox="0 0 24 24"
            >
              <path
                d="M3 17.25V21h3.75L17.81 9.94l-3.75-3.75L3 17.25zM20.71 7.04c.39-.39.39-1.02 0-1.41l-2.34-2.34a.9959.9959 0 0 0-1.41 0l-1.83 1.83 3.75 3.75 1.83-1.83z"
              />
            </svg>
          </div>
        </div>
        <div
          class="listgrid__list-item-item"
        >
          <svg
            aria-hidden="true"
            class="MuiSvgIcon-root MuiSvgIcon-colorError MuiSvgIcon-fontSizeMedium emotion-3"
            data-testid="GppBadIcon"
            focusable="false"
            viewBox="0 0 24 24"
          >
            <path
              d="M12 2 4 5v6.09c0 5.05 3.41 9.76 8 10.91 4.59-1.15 8-5.86 8-10.91V5l-8-3zm3.5 12.09-1.41 1.41L12 13.42 9.91 15.5 8.5 14.09 10.59 12 8.5 9.91 9.91 8.5 12 10.59l2.09-2.09 1.41 1.41L13.42 12l2.08 2.09z"
            />
          </svg>
        </div>
        <div
          class="listgrid__list-item-item"
        >
          <span
            class="listgrid__download"
          >
            <a
              href="https://opendata.arcgis.com/datasets/4c209944e2984b4a908a14b0cbe48075_0.zip"
              rel="noreferrer"
              target="_blank"
            >
              Rochester
            </a>
          </span>
        </div>
        <div
          class="listgrid__list-item-item"
        >
          USA
        </div>
        <div
          class="listgrid__list-item-item"
        >
          NY
        </div>
        <div
          class="listgrid__list-item-item"
        >
          <a
            href="mailto:null?subject=Water%20the%20Trees%20Source%20update%20request&body=Hi%2C%0A%20The%20tree%20source%20link%20for%20your%20city%20is%20broken%2C%20can%20you%20please%20update%20it%20here%20https%253A%252F%252Fwaterthetrees.com%252Fdata"
            rel="noreferrer"
            target="_blank"
          />
        </div>
        <div
          class="listgrid__list-item-item"
        />
        <div
          class="listgrid__list-item-item"
        />
        <div
          class="listgrid__list-item-item"
        />
      </div>
    </div>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
